compiler/rustc_trait_selection/src/traits/select/mod.rs

```diff
@@ -2417,13 +2417,18 @@ impl<'tcx> SelectionContext<'_, 'tcx> {
             }
 
             ty::Alias(ty::Opaque, ty::AliasTy { def_id, args, .. }) => {
-                // We can resolve the `impl Trait` to its concrete type,
-                // which enforces a DAG between the functions requiring
-                // the auto trait bounds in question.
-                match self.tcx().type_of_opaque(def_id) {
-                    Ok(ty) => t.rebind(vec![ty.instantiate(self.tcx(), args)]),
-                    Err(_) => {
-                        return Err(SelectionError::OpaqueTypeAutoTraitLeakageUnknown(def_id));
+                if self.infcx.can_define_opaque_ty(def_id) {
```

Comment:
this applies to all auto traits, not just `Freeze`, so it can impact selection in an incomplete way. I feel a bit uncomfortable about this, as it will break in the new solver due to ambiguity. Can you explicitly limit this to `Freeze` and add the following as a test
```rust
fn is_trait<T: Trait<U>, U: Default>(_: T) -> U { Default::default() }

trait Trait<T> {}
impl<T: Send> Trait<u32> for T {}
impl<T> Trait<i32> for T {}
fn foo() -> impl Sized {
    if false {
        is_trait(foo())
    } else {
        Default::default()
    }
}
```

Reply:
cc https://github.com/rust-lang/rust/pull/122192#issuecomment-2077466224****

Reply:
Hmm... we just recently made `Freeze` available for user defined bounds: https://github.com/rust-lang/rust/issues/121675

Reply:
hopefully the new solver is stable before freeeze is :3

alternatively: can you change this to force ambiguity instead?

Reply:
I did a thing, but it seems incomplete, I need to do some log analysis

Reply:
can't do that in confirmnation, you have to change `assemble_candidates_from_auto_impls` instead

Reply:
argml right. Old solver -.-